Multi-robot support (#349)
kroshu/kuka_drivers: doc/wiki/Home.md

## Multi-robot scenario

Since ROS 2 Jazzy, `ros2_control` supports asynchronous hardware interfaces. With this feature enabled, multiple robots can be started within the same `controller_manager`, because each hardware interface can run in its own asynchronous execution context, the blocking `read()` methods no longer cause an issue.

> For a multi-robot setup, a dedicated robot description xacro should be created that loads both robot models in one file using launch arguments (for example robot model names, prefixes and optional namespace-specific parameters). This combined xacro is then passed as the single `robot_description` to the control node. It is important that the synchronous hardware interface is activated last, because all component activation steps acquire a lock that blocks the read-write loop of the synchronous thread, therefore activation would starve out the already active harware interface of the main thread.

Reviewer: ```suggestion
For a multi-robot setup, a dedicated robot description xacro should be created that loads all robot models in one file using launch arguments (for example robot model names, prefixes and optional namespace-specific parameters). This combined xacro is then passed as the single `robot_description` to the control node. It is important that the synchronous hardware interface is activated last, because all component activation steps acquire a lock that blocks the read-write loop of the synchronous thread, therefore activation would starve out the already active hardware interface of the main thread.
```

> A dedicated launch file is also required for this setup. It should declare the arguments of both robots, generate the combined xacro, start one `controller_manager`, and spawn the controllers for both robots with the correct names and configuration files. Configuration files still need to be adapted to the corresponding namespaces and prefixed joint names. An example setup is available in the `kuka_multi_robot_examples` package: [examples/kuka_multi_robot_examples](https://github.com/kroshu/examples/tree/master/kuka_multi_robot_examples).

Reviewer: ```suggestion
A dedicated launch file is also required for this setup. It should declare the arguments of all robots, generate the combined xacro, start one `controller_manager`, and spawn the controllers for all robots with the correct names and configuration files. Configuration files still need to be adapted to the corresponding namespaces and prefixed joint names. An example setup is available in the `kuka_multi_robot_examples` package: [examples/kuka_multi_robot_examples](https://github.com/kroshu/examples/tree/master/kuka_multi_robot_examples).
```

The robot hardware desciptions expose two configurable parameters to control the async execution behavior:

- `async_thread_priority` (default: `69`): sets the thread priority for the asynchronous hardware interface executor thread
- `async_affinity` (default: `""` - empty, allows any core): pins the asynchronous hardware interface thread to specific CPU cores

> To plan with Moveit and a dual-arm setup, the moveit configuration also has to be modified. As here the URDF and SRDF files are not in the moveit support package, using MoveitConfigsBuilder is not recommended, the configuration files have to be loaded manually one by one. It is possible to create new configuration files with the resource names updated, or to remap the existing resource names from the launch files. An example for this second approach is also available in the `kuka_multi_robot_examples` package.

Reviewer: ```suggestion
To plan with Moveit and a multi-arm setup, the moveit configuration also has to be modified. As here the URDF and SRDF files are not in the moveit support package, using MoveitConfigsBuilder is not recommended, the configuration files have to be loaded manually one by one. It is possible to create new configuration files with the resource names updated, or to remap the existing resource names from the launch files. An example for this second approach is also available in the `kuka_multi_robot_examples` package.
```

> ### Dual-arm timing scenarios

Reviewer: ```suggestion
### Multi-arm timing scenarios
```

Author: The scenarios below only show a dual-arm setup, but I will write a note, that the same/similar holds for multi-arm as well

The following timing constraints apply in all cases due to `ros2_control` behavior:
- Main-thread `read` starts immediately after `write` finishes, so this thread is not idle.
- Async-thread `read` is called at a fixed rate (defined by the controller manager update rate), so there is an idle period after every `write`.
- KRCs send motion states every 4 ms, but jitter is possible.
- `update` runs only on the main thread, but it also updates the async hardware interface.

Note: for simplicity, in cases where it does not affect the outcome, `read` is triggered at the same time for both threads.

**Scenario 1:**
The async thread receives robot state 2 ms after `read` is triggered.
Outcome: both robots can be controlled smoothly, and jitter does not affect stability.

> ![alt text](resources/dual_arm_timing/scenario1.png)

Reviewer: Maybe also add a short list about the letters in the images (R &ndash; Read, W &ndash; write etc.).

> ![alt text](resources/dual_arm_timing/scenario1.png)

Reviewer: I think that GitHub supports PlantUML now. I believe you used that for these images.

Author: are you sure? I did not try it, but read that puml is not supported

Reviewer: Nope. I was referring to mermaid. But it's okay as is.

> **Scenario 2:**

Reviewer: Isn't this the non-problematic version of scenario 3 & 4? If so, adding a little note might be beneficial for better understanding.

The async thread receives robot state 0.5 ms after `read` is triggered.
Outcome: both robots can be controlled smoothly, but the system is not jitter-resistant.
![alt text](resources/dual_arm_timing/scenario2.png)

**Scenario 3:**
The async thread receives robot state 0.5 ms after `read` is triggered, and `read` scheduling is delayed for one cycle.

> Issue: the packet arrives while the thread is still idle. `read` is then called afterwards and skips this packet (which also causes a one-tick delay for all subsequent packets).

Reviewer: > which also causes a one-tick delay for all subsequent packets

I am not sure what this means. Will this stop RSI after 100 ticks, or does it simply add one missed packet to the statistics?

![alt text](resources/dual_arm_timing/scenario3.png)

**Scenario 4:**
The async thread receives robot state 0.5 ms after `read` is triggered. Main-thread `update` starts 0.5 ms after the state is received on the async thread. One packet is 0.5 ms late.

> Issue: due to the late packet, `update` is executed for the second time before this `write`. In the next cycle, no `update` is executed before `write`, causing a robot jerk.

Reviewer: ```suggestion
Issue: the late packet causes an extra `update` execution before `write`. Consequently, one set of commands produced by update is never transmitted via `write`. In the subsequent cycle, `write` sends stale data without an intervening `update`, causing a robot jerk.
```
